engine/src/main/java/org/hibernate/search/backend/impl/DeleteByQuerySupport.java

```diff
@@ -41,221 +29,55 @@ private DeleteByQuerySupport() {
-						}
-					}
-					else {
-						Type type = query.getType();
-						BytesRef valueAsBytes;
-						switch ( type ) {
-							case INT:
-							case FLOAT: {
-								int value;
-								if ( type == Type.FLOAT ) {
-									value = NumericUtils.floatToSortableInt( (Float) query.getValue() );
-								}
-								else {
-									value = (Integer) query.getValue();
-								}
-								BytesRefBuilder builder = new BytesRefBuilder();
-								NumericUtils.intToPrefixCoded( value, 0, builder );
-								valueAsBytes = builder.get();
-								break;
-							}
-							case LONG:
-							case DOUBLE: {
-								long value;
-								if ( type == Type.DOUBLE ) {
-									value = NumericUtils.doubleToSortableLong( (Double) query.getValue() );
-								}
-								else {
-									value = (Long) query.getValue();
-								}
-								BytesRefBuilder builder = new BytesRefBuilder();
-								NumericUtils.longToPrefixCoded( value, 0, builder );
-								valueAsBytes = builder.get();
-								break;
-							}
-							default:
-								throw new AssertionFailure( "has to be a Numeric Type at this point!" );
-						}
-						return new TermQuery( new Term( query.getFieldName(), valueAsBytes ) );
+					map.put( SingularTermDeletionQuery.QUERY_KEY, new SingularTermDeletionQueryMapper() );
+
+					DeletionQueryMapper[] arr = new DeletionQueryMapper[map.size()];
+					for ( Map.Entry<Integer, DeletionQueryMapper> entry : map.entrySet() ) {
+						arr[entry.getKey()] = entry.getValue();
 					}
 
+					MAPPERS = arr;
 				}
-
-			} );
-
-			TO_LUCENE_QUERY_CONVERTER = new ToLuceneQuery[map.size()];
-			for ( Map.Entry<Integer, ToLuceneQuery> entry : map.entrySet() ) {
-				TO_LUCENE_QUERY_CONVERTER[entry.getKey()] = entry.getValue();
+			}
+			finally {
+				MAPPERS_LOCK.unlock();
 			}
 		}
+		return MAPPERS[queryKey];
 	}
 
-	public static ToLuceneQuery getToLuceneQuery(int queryKey) {
-		return TO_LUCENE_QUERY_CONVERTER[queryKey];
-	}
-
-	private static final Set<Class<? extends DeletionQuery>> SUPPORTED_TYPES;
-	static {
-		{
-			Map<Integer, Class<? extends DeletionQuery>> map = new HashMap<>();
-
-			map.put( SingularTermQuery.QUERY_KEY, SingularTermQuery.class );
-
-			SUPPORTED_TYPES = Collections.unmodifiableSet( new HashSet<>( map.values() ) );
-		}
-	}
+	private static final Lock SUPPORTED_TYPES_LOCK = new ReentrantLock();
+	private static Set<Class<? extends DeletionQuery>> SUPPORTED_TYPES;
 
 	public static boolean isSupported(Class<? extends DeletionQuery> type) {
-		return SUPPORTED_TYPES.contains( type );
-	}
-
-	private static final StringToQueryMapper[] FROM_STRING;
-	static {
-		{
-			Map<Integer, StringToQueryMapper> map = new HashMap<>();
-
-			map.put( SingularTermQuery.QUERY_KEY, new StringToQueryMapper() {
+		if ( SUPPORTED_TYPES == null ) {
+			//LAZY init, but don't to it more than once
+			SUPPORTED_TYPES_LOCK.lock();
+			try {
+				if ( SUPPORTED_TYPES == null ) {
+					Map<Integer, Class<? extends DeletionQuery>> map = new HashMap<>();
```

Comment:
> He was talking about wrong versions of Lucene being loaded

I don't know how this would be easier or harder to detect depending on
whether this eagerly or lazily initialized? This class does not even import
any Lucene types at all?